all responces compiled (#1)
The-Marcy-Lab-School-Assignments/swe-sr-5-1-LAbreu00: README.md

```js
const friendsManager = {
  friends: [],
  addFriend(newFriend) {
    if (typeof newFriend !== "string") return;
    this.friends.push(newFriend);
  },
};

friendsManager.addFriend("daniel");
friendsManager.addFriend(true);
friendsManager.friends.push("emmaneul");
friendsManager.friends.push(42);
```

Explain how the code is not consistent or predictable, then provide an example in code that uses closure to make it more consistent and predictable.

### Response 2

> The reason this code is neither consistent, nor predictable is that there is a method inside of our object that takes an argument and if that argument's type is not a string then it stops execution, else if it is a string then it pushes the argument passed to method into friends array, But currently we could also push elements simply through the friends property and if a user was to do it through the friends property and not through the method then we could easily make the mistake of pushing an element that is not a string and we don't want that.
> 
> Instead we can use closure to keep our array protected from being mutated by any other means but the addFriend function. We would accomplish this by using encapsulation, which will require us to return an object only containing the addFriend method which references the friends array that in this example will be located in the surrounding scope of the friendsManager object. Since we're using closure now it means that the only way to gain access to our friends array will be through the object so we should also incorporate another method inside of the object friendsManager that returns a shallow copy of the array through destructuring.

Reviewer: This is excellent! Your response is really thorough. I especially like how you explain that the first example opens us up to making a mistake. This provides a clear "WHY" for making your suggested improvement.

Below, I've added some suggestions for improvement that I want you to consider! Your writing is excellent but this will take it to the next level.

* Note in my suggestion how replacing "the method" with "the `addFriend` method" removes any confusion. 
* Also note how the added commas and punctuation break up the sentences more cleanly. 
* In the first paragraph, explaining how the `addFriend` method works at a higher level gets the same point across without needing to explain the exact implementation.
* The example makes a copy using the spread operator, not destructuring

```suggestion
The reason this code is neither consistent nor predictable is that, while there is a `addFriend` method inside of our object that ensures that only strings are added to the `friends` array, we could also push elements simply through the `friends` property. If a user was to add a value directly through the `friends` property and not through the `addFriend` method, then we could easily make the mistake of pushing an element that is not a string. We don't want that.

Instead we can use closure to keep our array protected from being mutated by any other means other than the `addFriend` method. We would accomplish this by using encapsulation, which will require us to return an object only containing the `addFriend` method, which references the `friends` array that in this example, is located in the surrounding scope of the `getFriendsManager` function. Since we're using closure now, it means that the only way to gain access to our `friends` array will be through the `addFriend` method. We should also incorporate another method inside of the object `friendsManager` that returns a shallow copy of the array using the spread operator.
```

```javascript
const getFriendsManager = () => {
  const friends = [];
  const friendsManager = {
    addFriend(newFriend) {
      if (typeof newFriend !== "string") return;
      friends.push(newFriend);
    },
    getFriends() {
      return [...friends];
    },
  };
  return friendsManager;
};
```

## Prompt 3

With OOP in JavaScript, it's possible to use factory functions to achieve encapsulation and re-use them to make objects that look alike. However, factory functions have drawbacks and we often use classes instead.

How would you explain to a budding developer what the drawbacks of using factory functions are and why it is better to use classes instead?

### Response 3

While encapsulation can be done with **factory functions**, **classes** streamline the process.

- **Factory functions** create _copies_ of the original function every time it's called, on a small scale this isn't that big of a deal, but when you have _hundreds_ of functions it can start to take up a lot of _unnecessary_ space.

> - **Classes** on the other hand conserve that space and use the same function every time they're called. Additionally, **Classes** are all _self-contained_, making it easier to stay _organized_ no matter the size of the project.

Reviewer: This is a solid response. You identify the key difference which is that factory functions take up more space due to the fact that methods of the returned object are re-created each time the factory function is invoked.

However, I think someone who is just beginning to understand factory functions and classes would need more information to understand what you mean. When you say that factory functions "create copies of the original function", it is not clear what "the original function" is referring to? Are you referring to the factory function? 

It is never made clear that the object returned by the factory function would have methods and THOSE methods are the ones that need to be recreated each time. While you could explain this in words, a code snippet would do wonders here.

```js
const makePerson = (name) => {
  return {
    name,
    greet: () => console.log(`hello, my name is ${this.name}`)
  }
}

// vs.

class Person {
  constructor(name) {
    this.name = name;
  }
  greet() {
    console.log(`hello, my name is ${this.name}`)
  }
}
```

In this example, the `greet` method is recreated each time `makePerson` is invoked but is only created once for the `Person` class.

Can you come up with your own simple example?

## Prompt 4

Do some research on the history of when / how classes were introduced into JavaScript and share your findings. Your response should include:

- What version of JavaScript were classes introduced in and when did it come out?
- Why were classes introduced into JavaScript?

### Response 4

> Classes were introduce in the ECMAScript 6 (ES6), which came out in June 2015.
> Prior to ES6 JavaScript didn't have a formal class syntax and "classes" were simulated using constructor functions and prototypes which worked fine but didn't provide the clear structure that the class syntax provides. The concept of inheritance was too abstract prior to the ES6 update, and people migrating from other programming languages like Java or C# struggled with the level of abstraction before the update.
> 
> Simple additions like the extend keyword, which allows developers to create a relationship between two classes and the super method which allows users to extend the behavior of the parent class through one of its instances, in other words it allows us to call the constructor function or methods found in the parent class.

Reviewer: This is wonderful! Just adding in some backticks to highlight the `extends` and `super` keywords would make this perfect. Also, the last sentence is a bit long so adding in a period to break it up would be great.

```suggestion
Classes were introduce in the ECMAScript 6 (ES6), which came out in June 2015.
Prior to ES6 JavaScript didn't have a formal class syntax and "classes" were simulated using constructor functions and prototypes which worked fine but didn't provide the clear structure that the class syntax provides. The concept of inheritance was too abstract prior to the ES6 update, and people migrating from other programming languages like Java or C# struggled with the level of abstraction before the update.

Simple additions like the `extends` keyword, which allows developers to create a relationship between two classes and the `super` method which allows users to extend the behavior of the parent class through one of its instances. In other words it allows us to call the constructor function or methods found in the parent class.
```

## Prompt 5

OOP can still be achieved in JavaScript without using the `class` keyword and instead using the "Constructor Functions" and the "Prototype Chain" (look them up!)

```js
function Person(name, age) {
  this.name = name;
  this.age = age;
}

Person.prototype.greet = function () {
  return `Hi, I'm ${this.name}, and I'm ${this.age} years old.`;
};

const alice = new Person("Alice", 30);
console.log(alice.greet());
```

Provide one point that advocates for the use of this syntax and then provide a counter-argument for the use of classes instead.

### Response 5

> - Using "Constructor Functions" and the "Prototype Chain" syntax instead of **class** syntax can help deconstruct the mechanisms behind how classes work, providing a more in-depth understanding.
> - **Class** syntax provides some _syntactic sugar_, which can help make code both easier to work with and readable.

Reviewer: Nice response! it is really direct. Some suggestions on how you use quotes, bold, italic, and backticks. The rules I follow are:
* Use quotes to identify lingo, that is, terms used in the industry but that are distinctly different from technical concepts that you may want your reader to look up. The example here is "syntactic sugar". In the question we quoted "constructor functions" and the "prototype chain" which is appropriate since we wanted you to look those terms up.
* Use backticks to highlight code. So `class` in this case is a keyword that can be highlighted with backticks.
* Use bold to highlight key terms. While it does make sense to bold "class" here, I think its actually better to use backticks. You could bold "constructor functions" and "prototype chain" if you'd like to.
* Use italics to apply emphasis, not to highlight key words or phrases. Example: Using the `class` syntax results in a _much_ more readable programming experience.

```suggestion
- Using constructor functions and the prototype chain syntax instead of `class` syntax can help deconstruct the mechanisms behind how classes work, providing a more in-depth understanding.
- `class` syntax provides some "syntactic sugar", which can help make code both easier to work with and more readable.
```
